### PR TITLE
Add schizo/singularity component

### PR DESCRIPTION
**Pull request by rhc54**
Restore the old schizo/singularity component in place of the new ODLS
one so we can support both Singularity and non-Singularity jobs with the
same PRRTE instance. Merge the updates from the ODLS component into the
prior logic.

Signed-off-by: Ralph Castain <rhc@pmix.org>

### CODE REVIEW

**Comment by rhc54** [comment]
@gvallee Please give this a try. I haven't tested it, but attempted to correctly merge what you had done into the prior logic. Feel free to push changes to this PR and I'll commit when you feel it is ready.

**Comment by gvallee** [comment]
@rhc54 testing now

**Comment by gvallee** [comment]
@rhc54 first of all, thanks for doing this. After a quick test, here is what I see (I can work on this a little later, I need to switch to something else for a few days but can allocated time to this after that):
- specifying `--with-singularity` at configure time is I think not required. Singularity can be installed purely in user space now and using very different configurations so I would argue that it is better to discover the path at runtime. It would fairly easy to do, maybe something like use `--with-singularity` to compile and install the component and then use an implementation that would just detect where singularity is.
- the detection of the image format does not seem to be working:
```bash
$ prun -np 1 -x SY_EXEC_ARGS singularity exec  ~/.sympi/mpi_container_ubuntu-disco-openmpi-4.0.2-helloworld-hybrid/ubuntu-disco-openmpi-4.0.2-helloworld-hybrid.sif /opt/mpitest
```
```
[pessoa3:05087] mca:schizo:select: Querying component [slurm]
[pessoa3:05087] mca:schizo:select: Skipping component [slurm]. Query failed to return a module
[pessoa3:05087] mca:schizo:select: checking available component singularity
[pessoa3:05087] mca:schizo:select: Querying component [singularity]
[pessoa3:05087] mca:schizo:select: checking available component ompi
[pessoa3:05087] mca:schizo:select: Querying component [ompi]
[pessoa3:05087] mca:schizo:select: checking available component orte
[pessoa3:05087] mca:schizo:select: Querying component [orte]
[pessoa3:05087] mca:schizo:select: Skipping component [orte]. Query failed to return a module
[pessoa3:05087] mca:schizo:select: checking available component flux
[pessoa3:05087] mca:schizo:select: Querying component [flux]
[pessoa3:05087] mca:schizo:select: Skipping component [flux]. Query failed to return a module
[pessoa3:05087] Final schizo priorities
[pessoa3:05087] 	Schizo: ompi Priority: 10
[pessoa3:05087] 	Schizo: singularity Priority: 9
[pessoa3:05087] [[INVALID],INVALID] schizo:ompi: define_cli
/usr/bin/env: ‘run-singularity’: No such file or directory
```
- when specifying the singularity command on the `prun` command I am getting:
```bash
prun -np 1 -x SY_EXEC_ARGS singularity exec  ~/.sympi/mpi_container_ubuntu-disco-openmpi-4.0.2-helloworld-hybrid/ubuntu-disco-openmpi-4.0.2-helloworld-hybrid.sif /opt/mpitest
```
```
[pessoa3:05076] mca:schizo:select: Querying component [slurm]
[pessoa3:05076] mca:schizo:select: Skipping component [slurm]. Query failed to return a module
[pessoa3:05076] mca:schizo:select: checking available component singularity
[pessoa3:05076] mca:schizo:select: Querying component [singularity]
[pessoa3:05076] mca:schizo:select: checking available component ompi
[pessoa3:05076] mca:schizo:select: Querying component [ompi]
[pessoa3:05076] mca:schizo:select: checking available component orte
[pessoa3:05076] mca:schizo:select: Querying component [orte]
[pessoa3:05076] mca:schizo:select: Skipping component [orte]. Query failed to return a module
[pessoa3:05076] mca:schizo:select: checking available component flux
[pessoa3:05076] mca:schizo:select: Querying component [flux]
[pessoa3:05076] mca:schizo:select: Skipping component [flux]. Query failed to return a module
[pessoa3:05076] Final schizo priorities
[pessoa3:05076] 	Schizo: ompi Priority: 10
[pessoa3:05076] 	Schizo: singularity Priority: 9
[pessoa3:05076] [[INVALID],INVALID] schizo:ompi: define_cli
--------------------------------------------------------------------------
prte was unable to find the specified executable file, and therefore
did not launch the job.  This error was first reported for process
rank 0; it may have occurred for other processes as well.

NOTE: A common cause for this error is misspelling a prte command
      line parameter option (remember that prte interprets the first
      unrecognized command line token as the executable).

Node:       pessoa3
Executable: singularity
--------------------------------------------------------------------------
```

In summary: **I think this is the way to go** (it is in bold because I find quite exciting) and after reading the code and looking at the errors, everything seems to be minor and easy to fix/adapt. I will allocate time to work on this as soon as possible but I unfortunately need to switch to something else this week.

Thanks again!

**Comment by rhc54** [comment]
> specifying --with-singularity at configure time is I think not required. Singularity can be installed purely in user space now and using very different configurations so I would argue that it is better to discover the path at runtime. It would fairly easy to do, maybe something like use --with-singularity to compile and install the component and then use an implementation that would just detect where singularity is.

My thought was that the ```--with-singularity``` option could be used to specify the default install of Singularity on the machine. The configure logic will look to see if Singularity is in the default path or in the specified one - if the user then doesn't provide a path in the environment, this path is used.

Note that the schizo/singularity component is built regardless of whether or not ```singularity``` is found in the default path, so the component will still activate if the user specifies a location. Configure _will_ exit, however, if the person building PRRTE specifies a location for Singularity and we don't find it. This is consistent with our configure/build policy since we cannot do what the person building PRRTE has requested.

Does that make sense?

**Comment by rhc54** [comment]
> the detection of the image format does not seem to be working:

Ah - I only looked for ```.sif``` in argv[0]. I should perhaps have looked for ```singularity``` in argv[0] and for ```.sif``` in the appropriate location. I'll let you determine the best logic.

**Comment by rhc54** [comment]
> when specifying the singularity command on the prun command 

Ouch - I automatically added ```singularity exec``` to the front of the argv for the app, but you already have it there since you included those params in the ```prun``` command. Probably need to check for them and avoid duplication.

**Comment by gvallee** [comment]
> > specifying --with-singularity at configure time is I think not required. Singularity can be installed purely in user space now and using very different configurations so I would argue that it is better to discover the path at runtime. It would fairly easy to do, maybe something like use --with-singularity to compile and install the component and then use an implementation that would just detect where singularity is.
> 
> My thought was that the `--with-singularity` option could be used to specify the default install of Singularity on the machine. The configure logic will look to see if Singularity is in the default path or in the specified one - if the user then doesn't provide a path in the environment, this path is used.
> 
> Note that the schizo/singularity component is built regardless of whether or not `singularity` is found in the default path, so the component will still activate if the user specifies a location. Configure _will_ exit, however, if the person building PRRTE specifies a location for Singularity and we don't find it. This is consistent with our configure/build policy since we cannot do what the person building PRRTE has requested.
> 
> Does that make sense?

If the component is compiled and installed regardless of the `--with-signularity`, it is all good. With my stack, there is no default Singularity, you can install singularity as a user, after the installation of the entire stack and even switch back and forth between versions. 

So yes it makes sense and it sounds good to me.

**Comment by gvallee** [comment]
@rhc54 I just created a PR against your branch. With it, I can run everything I have in mind and could not find a use case that would not work. Let me know what you think about it.
